### PR TITLE
Replace faster-coco-eval with ultrafast-pycocotools

### LINKED ISSUE #26095: Consider an opt-in ultrafast-pycocotools backend for COCO evaluation
### Search before asking

- [x] I searched existing issues and pull requests for this backend and found no matching feature request.

### Description

Would you consider an opt-in `ultrafast-pycocotools` evaluator for the existing COCO evaluation path? I maintain the library and am requesting agreement on the approach before submitting a feature PR, as required by the contribution guidelines.

The library is a BSD-2-Clause Rust/Python COCO evaluator, [available on PyPI](https://pypi.org/project/ultrafast-pycocotools/0.1.5/):

```bash
python -m pip install ultrafast-pycocotools==0.1.5
```

Version 0.1.5 provides tested CPython 3.9–3.14 wheels for Linux x86-64, Windows x86-64, and macOS Intel/Apple Silicon. Ultralytics' existing Python 3.8 support should remain on the current backend; this proposal would not raise the core Python requirement or add a mandatory dependency.

Version 0.1.5 fixes source-archive license packaging; the evaluator implementation is unchanged from the measured 0.1.4.

### Use case

Reduce COCO evaluation time and peak host memory during validation, particularly for saved predictions and repeated large evaluations. I inspected `DetectionValidator.coco_evaluate` at revision `88b030173dbbaff2f22925faffd0cb780500c1a9`, which already uses faster-coco-eval and also serves segmentation and pose validation.

The proposed change would stay in this shared owner, keep faster-coco-eval as the default, and permit explicit selection of ultrafast. I would avoid a separate validator implementation and a new global CLI option unless maintainers prefer one. Existing metric keys, image selection, cached GT behavior, and fitness computation must remain unchanged. COCO detection/segmentation/pose and LVIS behavior would need focused validation before any PR is marked ready.

### Evidence and limits

A saved-prediction benchmark using YOLO26n on all 5,000 COCO val2017 images (596,202 detections) produced:

| Backend | Scoring time | Whole-process peak RSS |
|---|---:|---:|
| pycocotools 2.0.11 | 37.073 s | 1,599.2 MiB |
| faster-coco-eval 1.8.0 | 7.257 s | 1,651.6 MiB |
| ultrafast 0.1.2 | 1.942 s | 745.9 MiB |

This historical **0.1.2** experiment is a single fresh process per backend, not a repeated estimate for the newly published 0.1.5. It includes GT indexing/result loading/evaluation/accumulation/summary, excludes JSON parsing and inference, and includes parsed inputs in peak RSS. It does **not** measure the complete current Ultralytics validator. All ultrafast precision, recall, scores, and stats arrays match pycocotools byte for byte. [Recorded inputs, versions, timings, and output hashes](https://github.com/developer0hye/ultrafast-pycocotools/blob/main/bench/results/yolo26n_v012.json).

The [0.1.4 paired report](https://github.com/developer0hye/ultrafast-pycocotools/blob/main/docs/efficiency-v014.md) separately measures YOLO26n file evaluation at 0.628 s and 223.2 MiB median over three runs, including JSON loading. It compares 0.1.3 with 0.1.4; these different timing scopes should not be mixed to claim a speedup against the table above.

As additional framework evidence, [actual RF-DETR one-pass metric replay](https://github.com/developer0hye/ultrafast-pycocotools/blob/main/docs/rfdetr.md) with 0.1.4 improved from 14.130 s / 3,301.7 MiB to 6.747 s / 2,073.2 MiB versus faster-coco-eval, with exactly equal output metric tensors. That result belongs to RF-DETR and is not a claimed Ultralytics speedup.

### Additional

- [Library and reproducible benchmarks](https://github.com/developer0hye/ultrafast-pycocotools)
- [Compatibility and CI coverage](https://github.com/developer0hye/ultrafast-pycocotools/blob/main/docs/ci.md)
- The library exposes the metric-key aliases, logging callback, and LVIS mode used by the current evaluator. Full integration should still be assessed against real supported task paths and version floors, rather than inferred from API names alone.
- I would first agree on a minimal selection mechanism and validation requirements, then submit a small PR. Default replacement could be considered separately after broader validation.

### Are you willing to submit a PR?

- [x] Yes, I would like to submit a PR after approval of the approach.

### PR DESCRIPTION
**Pull request by developer0hye**
Replaces faster-coco-eval with ultrafast-pycocotools in the shared COCO/LVIS validator and its existing mask/annotation consumers. The runtime requires the public **0.1.7** release, including the supported CPython 3.8/macOS arm64 wheel. No backend selector, platform shim or fallback is added.

The three real-prediction smoke cases now use independent official COCO annotations and compare evaluator metrics/fitness and cached reuse. Five synthetic COCO/LVIS regressions remain. The regular CI matrix installs both evaluators so these tests execute.

Independent real-data validation on the i5-10400 / RTX 3070 host completed 94 processes. All returned metrics and fitness are exactly equal for all 5,000 COCO images in detection, segmentation and pose; complete arrays differ by at most 2.23e-16. Whole-validation medians improve 16.3%, 9.1% and 10.0%, respectively. Pose whole-validation peak RSS increases 0.9%; evaluator-only RSS decreases. The report also covers official LVIS bbox compatibility, an explicitly labelled additional mask subset, the original reference mask failure, cold/cached replay and background load.

- [Report and reproduction](https://github.com/developer0hye/ultrafast-pycocotools/blob/dfba260fcab6a9b2307b53b98f220e199dbe75e2/docs/ultralytics-pr26101-validation.md)
- [Item-by-item reviewer checklist](https://github.com/developer0hye/ultrafast-pycocotools/blob/dfba260fcab6a9b2307b53b98f220e199dbe75e2/docs/ultralytics-review-checklist.md)
- [Public wheel/native installation proof](https://github.com/developer0hye/ultrafast-pycocotools/blob/dfba260fcab6a9b2307b53b98f220e199dbe75e2/bench/results/release_v017_verification.json)
- [Actual predictions, annotations, full arrays and logs](https://github.com/developer0hye/ultrafast-pycocotools/releases/download/v0.1.7/ultralytics-pr26101-evidence-20260909.tar.gz)

Current head: 0451a3c9282461e74cbaeeb5d03991be9162cdaf. All five regular test targets pass (Linux x86/ARM, macOS ARM, Windows and Python 3.8). CI is not reported green: an ARM semantic benchmark failed downloading cityscapes8.zip with HTTP 500 and requires a maintainer rerun.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @developer0hye, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: pyproject.toml [hunk 1]
```diff
@@ -170,7 +170,7 @@ logging = [
 extra = [
     "ipython", # interactive notebook
     "albumentations>=1.4.6", # training augmentations
-    "faster-coco-eval>=1.6.7", # COCO mAP
+    "ultrafast-pycocotools>=0.1.6", # COCO mAP
```
Comment (UltralyticsAssistant) [inline pyproject.toml]: 💡 **MEDIUM**: This keeps Ultralytics' Python >=3.8 contract, but ultrafast-pycocotools 0.1.6 does not publish a CPython 3.8 macOS arm64 wheel and requires a Rust source build on that platform. Consequently, `pip install ultralytics[extra]` fails for supported Python 3.8/macOS arm64 users without a Rust toolchain, and the runtime `check_requirements()` path has the same problem. Please publish the missing wheel or retain a supported evaluator path for this platform. ([]())

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all 8 changed files, including validator control flow, dependency wiring, Objects365 conversion, Comet decoding, CI, and integration tests. The evaluator integration is otherwise consistent with the repository usage and added parity coverage; one supported-platform packaging regression remains.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `pyproject.toml:173` This keeps Ultralytics' Python >=3.8 contract, but ultrafast-pycocotools 0.1.6 does not publish a CPython 3.8 macOS arm64 wheel and requires a Rust source build on that platform. Consequently, `pip install ultralytics[extra]` fails for supported Python 3.8/macOS arm64 users without a Rust toolchain, and the runtime `check_requirements()` path has the same problem. Please publish the missing wheel or retain a supported evaluator path for this platform. ([]())
</details>

**Review by glenn-jocher — CHANGES_REQUESTED** [review]
The runtime replacement is appropriately centralized, and I independently ran the eight added evaluator/serialization cases: all passed on Python 3.13/macOS arm64. The parity coverage exercises real implementations and is justified by the scope of this dependency replacement.

One packaging blocker remains at the reviewed head. PyPI provides faster-coco-eval 1.6.7 and 1.7.2 CPython 3.8 macOS arm64 wheels, but ultrafast-pycocotools 0.1.6 has no wheel for that supported target. This changes installation from a binary wheel to a Rust source build. Please publish and validate the missing wheel, or a new stable release that provides it, and update the dependency floor here if needed. Do not add a fallback backend, selector argument, or platform shim.

This is the same unresolved packaging thread from the automated review. Once the supported wheel is available, the remaining review can be completed on the exact live head.

**Comment by glenn-jocher** [comment]
Additional acceptance requirement: this dependency replacement needs independent before/after accuracy and performance validation on identical real predictions for detection, instance segmentation, and pose, plus LVIS compatibility. The current synthetic parity cases are useful regression checks, but the end-to-end smoke cases construct ground truth from the predictions and only require AP to fall between zero and one; they cannot establish real-dataset accuracy preservation.

The published full-COCO timing is detection-only. Please provide repeatable task-by-task evidence with exact source/package revisions, dataset and prediction hashes, all returned metrics and fitness, maximum differences in complete precision/recall/score arrays, alternating fresh-process wall times, and process peak memory. Distinguish evaluator time from whole-validation time, and include cached-ground-truth repeated evaluation. Approval remains blocked on this evidence as well as the missing supported wheel. No backend selector, new user argument, or fallback shim.

**Comment by glenn-jocher** [comment]
Independent validation on ultra5 is complete for the reviewed head e600c6f2275acd4344127ea59942fccc179a7c2f against its exact parent 4e6701fe01eb49d719923029d8a30226d8a2c987.

Real full validation used YOLO26n / YOLO26n-seg / YOLO26n-pose, imgsz=640, batch=8, workers=4, FP32, seed=0, on an RTX PRO 6000 Blackwell. Environment: Python 3.10, PyTorch 2.7.0+CUDA 12.8, NumPy 1.26.4, faster-coco-eval 1.8.0, ultrafast-pycocotools 0.1.6. Existing training jobs stayed active.

| Task | Images | Predictions | Main mAP50–95, before = after | Cold evaluator median, before → after | Peak process RSS, before → after |
|---|---:|---:|---:|---:|---:|
| detect | 5,000 | 747,280 | 0.4088107889 | 7.005 s → 0.505 s | 2931 MiB → 1287 MiB |
| segment | 5,000 | 738,284 | 0.3443379136 | 17.219 s → 6.244 s | 4693 MiB → 3646 MiB |
| pose | 2,346 | 82,052 | 0.5669189584 | 2.586 s → 1.283 s | 1206 MiB → 1196 MiB |

For all three tasks, before/after prediction JSON hashes are identical and every returned full-validation metric, including fitness and box/mask/pose metrics, is exactly equal. Complete precision/recall/score arrays and evaluator summaries differ by at most 2.23e-16. Every cold and cached-ground-truth replay agrees with its corresponding full-validation metrics within 1e-12.

Timings are three alternating fresh-process samples per backend, with the real validation dataset/dataloader determining image IDs, predictions retained in jdict, OMP_NUM_THREADS=2, RAYON_NUM_THREADS=2, and OPENBLAS_NUM_THREADS=1. Each process ran a cold and a cached-ground-truth call. Inference, initial prediction loading, and dataset setup are outside the evaluator clock. These are evaluator speedups, not claimed whole-validation/training speedups; the host was not otherwise isolated.

Additional checks: the official LVIS 100-image bbox example matches complete arrays within 2.23e-16; its published result file contains no segmentation predictions. The PR’s five COCO/LVIS edge-case parity tests and three serialization cases pass locally on Python 3.13/macOS arm64. One hundred actual COCO image/annotation lookups match through the APIs used by Objects365, and 100 real segmentation RLEs decode identically through the mask APIs used by Comet.

**Remaining blocker: the supported Python 3.8/macOS arm64 binary wheel.** I reproduced it with:

```bash
python -m pip download --no-deps --only-binary=:all: --platform macosx_11_0_arm64 --python-version 38 --implementation cp --abi cp38 "faster-coco-eval>=1.6.7"
# succeeds: faster_coco_eval-1.7.2-cp38-cp38-macosx_11_0_arm64.whl
python -m pip download --no-deps --only-binary=:all: --platform macosx_11_0_arm64 --python-version 38 --implementation cp --abi cp38 "ultrafast-pycocotools>=0.1.6"
# fails: no matching distribution
```

The measured runtime benefit is real across these tested tasks. Changes requested remains appropriate until the missing supported wheel is published and validated, with a dependency-floor update here if a new release is needed. No selector argument, fallback backend, or platform shim.

**Comment by developer0hye** [comment]
@glenn-jocher The missing wheel and dependency floor are fixed at **0451a3c9282461e74cbaeeb5d03991be9162cdaf**. Thanks for the independent validation in your follow-up. Your packaging and prediction-derived-GT findings were valid; both have concrete fixes now.

**ultrafast-pycocotools 0.1.7 is public on PyPI**, including `cp38-cp38-macosx_11_0_arm64`. A fresh native M2 installation using `--only-binary=:all:` passes **249 tests** (17 optional skips). All 35 wheels plus the sdist match the tested publication artifacts. [Release and installation verification](https://github.com/developer0hye/ultrafast-pycocotools/blob/dfba260fcab6a9b2307b53b98f220e199dbe75e2/bench/results/release_v017_verification.json).

I also completed a separate **94-process** before/after run on the i5-10400 / RTX 3070 host: all 5,000 COCO val2017 images for detect, segment and pose, FP32/640/batch 16/workers 2, Python 3.12.3, Torch 2.13.0+cu126, NumPy 2.4.4, faster-coco-eval 1.8.0 versus ultrafast 0.1.7. The measured replacement is 0d406853; the only later change at 0451a3c9 installs both evaluators in the regular CI matrix. Runtime/test sources are identical. These are separate predictions/settings from your ultra5 run.

| Task | Main AP, before = after | Whole validation median | Cold evaluator median | Replay peak RSS |
| --- | ---: | ---: | ---: | ---: |
| detect | 0.4081073392 | 37.782 → 31.609 s | 7.681 → 0.849 s | 2887.0 → 1370.3 MiB |
| segment | 0.3436595381 | 153.892 → 139.950 s | 21.844 → 5.836 s | 4293.2 → 3789.2 MiB |
| pose | 0.5638424803 | 35.915 → 32.330 s | 5.251 → 1.817 s | 1794.4 → 1719.8 MiB |

All returned full-validation metrics and fitness are exactly equal. Complete precision/recall/score arrays differ by at most **2.23e-16**; pose arrays are exactly equal. Cold and cached calls agree with the full-validation COCO metrics, and arrays are identical across cached reuse within each backend. Whole-validation peak RSS falls for detect/segment, but **pose increases from 2726.0 to 2750.7 MiB (+0.9%)**; the report retains that result. Other processes stayed active, all samples are retained, and the report separates evaluator, whole-validation and fresh-process time.

The complete [report](https://github.com/developer0hye/ultrafast-pycocotools/blob/05e51ad5799af358b6c45d4f431a6be550650470/docs/ultralytics-pr26101-validation.md), [raw metric/telemetry JSON](https://github.com/developer0hye/ultrafast-pycocotools/blob/dfba260fcab6a9b2307b53b98f220e199dbe75e2/bench/results/ultralytics_pr26101_20260909.json), and [269 MB evidence archive](https://github.com/developer0hye/ultrafast-pycocotools/releases/download/v0.1.7/ultralytics-pr26101-evidence-20260909.tar.gz) include the actual predictions, independent annotations, full arrays, exact executed scripts and test logs. All 245 archived file hashes were verified. Archive SHA-256: `8c9dd4d1cf3a0a6e1b4e83a49a7b799023756294a2108fc7389efffc5aafa33e`.

Item-by-item acceptance checklist ([expanded checklist with evidence links](https://github.com/developer0hye/ultrafast-pycocotools/blob/dfba260fcab6a9b2307b53b98f220e199dbe75e2/docs/ultralytics-review-checklist.md)):

- [x] **Supported wheel:** publish and natively validate the CPython 3.8/macOS arm64 binary.
- [x] **Dependency floor:** require >=0.1.7 in production and CI requirements.
- [x] **Independent smoke-test GT:** use original COCO annotations, not generated predictions; compare against the reference evaluator.
- [x] **Synthetic regressions:** retain all five COCO/LVIS edge cases; both packages are now installed in normal CI so the eight comparison cases execute.
- [x] **Real detection accuracy:** 5,000 images / 733,070 predictions; all metrics and fitness identical.
- [x] **Real segmentation accuracy:** 5,000 images / 724,953 predictions; all bbox/mask metrics and fitness identical.
- [x] **Real pose accuracy:** 5,000 images / 134,663 predictions; all bbox/keypoint metrics and fitness identical.
- [x] **LVIS compatibility:** official 100-image bbox input matches; additional 93-image annotated real-mask subset matches complete bbox/mask arrays. Details and limitations below.
- [x] **Exact revisions/packages:** source revisions, native hashes, frozen executed scripts and complete environment saved.
- [x] **Dataset/prediction hashes:** official source URLs, models, image hashes, manifests, annotations and identical prediction files saved.
- [x] **Every returned metric/fitness:** full dictionaries saved for every call; key sets and full/replay results checked.
- [x] **Complete arrays:** actual precision/recall/scores saved as NPZ, with shapes and maximum differences; diagnostic overhead excluded from timing.
- [x] **Alternating fresh processes:** four full-validation and six replay rounds per backend, balanced order, every sample retained.
- [x] **Peak memory/background load:** process RSS, sampled process-tree RSS, host CPU and RAM recorded; the pose whole-validation memory increase is explicit.
- [x] **Evaluator versus whole validation:** separate nested evaluator, full-validator and fresh-process clocks.
- [x] **Cached GT:** one cold plus two calls reusing the identical COCO object; metrics and full arrays checked after reuse.
- [x] **No selector/shim/fallback:** the runtime replacement stays in the shared validator owner; no new user argument.
- [ ] **Terminal-green live-head CI:** All five regular test targets pass: Linux x86, Linux ARM, macOS ARM, Windows and Python 3.8. One unrelated ARM benchmark download still needs a maintainer rerun, described below.

LVIS scope: the official 100-image prediction file has no masks. For additional mask coverage, actual YOLO26n-seg masks were mapped using LVIS's pinned official COCO-to-synset mapping. On the original 100-image input, faster-coco-eval 1.8.0 and 1.7.2 both fail in mask preparation (`unordered_map::at` / `_Map_base::at`); 0.1.7 completes it, including cached reuse. The archive preserves both failures. The separately labelled 93-image annotated subset is the paired mask comparison, with metric delta <=5.56e-17 and arrays <=2.23e-16. This is a compatibility workload, not a full-taxonomy LVIS model accuracy claim.

The remaining CI failure is [ARM Benchmark SemanticSegmentationModel](https://github.com/ultralytics/ultralytics/actions/runs/34245506499/job/102126320973): `cityscapes8.zip` returned HTTP 500, then validation failed because the downloaded dataset was missing. The file now responds successfully, but my account cannot rerun upstream jobs. The workflow has now finished; rerun fails with "Must have admin rights to Repository." Could you rerun that failed job? I have not labelled this run green.

Documentation clarification: COCO evaluation retains the existing defaults of `maxDets=100` for bbox/mask and `maxDets=20` for keypoints (the model prediction cap is 300). The linked report corrects the earlier prose; no measurements, arrays or package versions in this reply changed.

File: pyproject.toml (hunk 1 above)
Comment (developer0hye) [reply pyproject.toml]: Fixed in the public 0.1.7 release, and this PR now requires >=0.1.7. A fresh native CPython 3.8/macOS arm64 installation with --only-binary=:all: passes 249 tests (17 optional skips); the published wheel SHA-256 matches the tested release artifact. See the [public distribution/test proof](https://github.com/developer0hye/ultrafast-pycocotools/blob/dfba260fcab6a9b2307b53b98f220e199dbe75e2/bench/results/release_v017_verification.json) and the [item-by-item response](https://github.com/ultralytics/ultralytics/pull/26101#issuecomment-5588018429). No platform shim or fallback was added.

**Comment by glenn-jocher** [comment]
Thanks, the `0.1.7` wheel and dependency-floor updates address the remaining review blockers. I’ve confirmed the only failing check is the unrelated ARM `cityscapes8` download failure; please leave the head unchanged while we rerun that job from the maintainer side.

**Comment by developer0hye** [comment]
@glenn-jocher Thanks for confirming the wheel and dependency-floor blockers are addressed. The reviewed head remains **0451a3c9282461e74cbaeeb5d03991be9162cdaf**, as requested.

A dependency release update: **[ultrafast-pycocotools 0.1.11 is now public on PyPI](https://github.com/developer0hye/ultrafast-pycocotools/releases/tag/v0.1.11)**. It includes further pose loading/storage and numeric-decoding optimizations. The complete 35-wheel matrix, including CPython 3.8/macOS arm64, and the source archive passed their tests; all 36 public files match the tested publication checksums.

The [new comparison and audit checklist](https://github.com/developer0hye/ultrafast-pycocotools/blob/v0.1.11/docs/hotcoco-performance-goal.md) covers all 5,000 COCO images on M2 and i5-10400: the source build now shipped in 0.1.11 beat hotcoco 1.0.0 in median time and peak RSS for all 24 tested task/input/pool configurations, while retaining byte-identical pycocotools arrays. These are separate standalone evaluator measurements; the earlier Ultralytics/faster-coco-eval whole-validation results and their limitations remain unchanged.

0.1.11 satisfies this PR's existing `>=0.1.7` requirement. No commits were added. The ARM `cityscapes8` benchmark check still shows the previously reported failure, so I am continuing to leave the head unchanged for the maintainer-side rerun.

**Comment by glenn-jocher** [comment]
Merged `main` into this branch to clear the conflict. The only conflict was in `coco_evaluate`: #26246 rewrote the three size-mAP lines to write under the suffix of the IoU type being evaluated and to skip ranges the evaluator does not report, which already covers the `iou_type != "keypoints"` guard this PR added there, so I kept the `main` version of those lines. `test_ultrafast_pycocotools` and `test_coco_evaluator_parity` pass on the merged branch (8 passed, detect, segment and pose, with `ultrafast-pycocotools` 0.1.11 and `faster-coco-eval` as the reference).

**Comment by glenn-jocher** [comment]
@developer0hye thanks for the thorough work on this. I pushed `367169055`, which reduces the PR to a pure evaluator replacement: **+29/-49 over 10 files**, down from +1542/-41.

- **Removed `tests/fixtures/`** (the 1378-line `coco8-annotations.json` and its README). We do not commit dataset annotations to the repository.
- **Removed `test_coco_evaluator_parity` and the fixture-based test.** They were valuable as one-time evidence, and that evidence is recorded in this thread (identical mAP on full COCO for detect, segment and pose, and I ran all eight parity cases on the merged branch against `faster-coco-eval`, all passing to 1e-12). Keeping them would keep `faster-coco-eval` installed in CI permanently as a reference for a dependency we are removing.
- **Replaced the old smoke test rather than retargeting it.** Your fixture exposed a real gap: `test_faster_coco_eval` on `main` never evaluates anything. coco8 images come from train2017, so loading predictions against `instances_val2017.json` fails with `Results do not correspond to current coco set`, which `coco_evaluate` logs and swallows. The new `test_ultrafast_pycocotools` runs a detect `val` with `save_json=True`, which builds COCO-format ground truth from the dataset labels and runs the real evaluator, and asserts no `unable to run` warning was logged. I checked that it passes with `ultrafast-pycocotools` 0.1.11 and fails when the evaluator import is broken.
- **CI installs** now replace `faster-coco-eval` with `ultrafast-pycocotools` in the three jobs that had it, instead of installing both, and the floor job is back to what `main` has.
- **Docs:** two remaining `faster-coco-eval` mentions (`docs/en/modes/val.md`, `docs/macros/validation-args.md`) are updated.
- `lvis_protocol="coco"` stays, since the library default (`official`) would change the LVIS numbers, and the `COCOeval` call is compacted.
